### ISSUE TITLE
[Mobile] Model output is different in React Native compared to NodeJS

### ISSUE THREAD

**Issue by pax-k** [microsoft/onnxruntime#19353]
### Describe the issue

I'm using [Xenova/all-MiniLM-L6-v2](https://huggingface.co/Xenova/all-MiniLM-L6-v2) to extract embeddings from sentences. Given [this](https://gist.github.com/pax-k/63f42c74a9d39ea1f7cd0351b5267bec#file-onnx-minilm-ts-L95) inference code, I execute it as is in both NodeJS and React-Native (in RN with a [slight difference](https://github.com/microsoft/onnxruntime-inference-examples/blob/main/mobile/examples/React_Native/ort-rn-basic-usage/App.tsx#L17) in how loading the model is made).

The NodeJS outputs are good. The problem is that I get slightly different vector embeddings in React Native, using the same code.

Things to note from the inference code:
- the model was downloaded locally as seen in the imports
- uses BertTokenizer
- tokenizer inputs are padded and truncated
- no pooling, no normalization
- returns raw data from the model as is, without any transformations:
```
let output = await session.run(input);
return Array.from(output.last_hidden_state.data);
```

### To reproduce

Setup iOS:
```
git clone https://github.com/pax-k/expo-onnx
cd expo-onnx
yarn
npx expo prebuild
npx pod-install
yarn ios
// copy the sortedSentences JSON object from the console
```

Setup NodeJS:
```
git clone https://github.com/pax-k/nodejs-onnx
cd nodejs-onnx
yarn
npx ts-node --esm onnx-minilm.ts
// copy the sortedSentences JSON object from the console
```

I combined and analyzed the 2 JSONs in [this](https://stackblitz.com/edit/js-yjwqqm?file=index.js) project:
- open it
- check the browser console
- you can see this table:

<img width="1413" alt="Screenshot 2024-01-31 at 14 17 42" src="https://github.com/microsoft/onnxruntime/assets/23238077/f3550ca0-0b8d-4d47-9c2c-d3b1d9c475fa">

`a` is from NodeJS
`b` is from React Native

Observations:
- none of the NodeJS embeddings match any of the React Native embeddings
- manhattan distance should be zero
- cosine similarity is expected to be different in NodeJS vs React Native because dot product is prone to floating point errors (if you run [this](https://gist.github.com/pax-k/f84b2e6ae5e517a13140d010d2435bc5) code in both contexts, you will get different results)

### Urgency

It's pretty urgent 

### Platform

React Native

### OS Version

iOS 17.0.1, iPhone 15 Pro Simulator

### ONNX Runtime Installation

Released Package

### Package Name (if 'Released Package')

onnxruntime-react-native

### ONNX Runtime Version or Commit ID

1.16.3

### ONNX Runtime API

JavaScript

### Architecture

ARM64

### Execution Provider

Default CPU

**Comment by simonwh** [comment]
Just chiming in to say I can replicate this on my machine. I hope @skottmckay or someone else from the team can give a few hints to why this might be happening!

**Comment by skottmckay** [comment]
My initial thought based on looking at the [json output](https://stackblitz.com/edit/js-yjwqqm?file=mobile.json,node.json) is that this is pretty typical and expected when there are lots of floating-point operations being executed on different platforms. 

If you look at the model in Netron there are many MatMul operations. The order of the individual operations will affect the exact value produced by that node. There are [many additions and multiplications](https://en.wikipedia.org/wiki/Matrix_multiplication) in a single MatMul. The order of each set of multiplications matter. The order those are added together matter. But there's no rule about the order. i.e. mathematically a x b x c == c x b x a, but you'll get two different results due to how floating-point numbers work. 

The low-level instructions used to execute the operations differ by platform/architecture (e.g. various AVX instruction sets on intel/amd, NEON on arm, etc.). These differences accumulate with each node and are magnified by nodes that do a lot of calculation (e.g. MatMul/Conv/Gemm). 

When you say the NodeJS output is 'good' but the React Native ones aren't how is that assessed? By the output from using the embeddings in a downstream model? Or is it that the floating-point values differ beyond some expected tolerance?

**Comment by simonwh** [comment]
@skottmckay 

We’ve been comparing both the vectors produced by the model directly by loading it with sentence_transformer in python vs. ONNX in node and here we see the exact same results. That’s why we say they look “correct”.

We then compared with ONNX in node vs. react-native and saw widely different results for some inputs.

We understand that floating point calculations can vary slightly on different architectures, but we didn’t expect to see discrepancies this big and seemingly random.

In the comparison chart, we show the Manhattan distance between vectors produced by ONNX node vs ONNX react-native. For most, you can see that the difference is really small (< 0.001), and expected due to differences in floating point math.

However, for a few vectors you will see a huge difference, eg.:

- “Andrei works at Kin”
- “Christopher works at-> Kin”
- “Me co-founded Kin”
- “0.1.1 version of Kin” (while “0.1.4 version of Kin” is almost identical)
- “Me looks for in candidates align with Kin’s values”

What can be the contributing factor to these difference? The inconsistency creates a blocker for us to put this in production.

**Comment by pax-k** [comment]
@skottmckay

> When you say the NodeJS output is 'good' but the React Native ones aren't how is that assessed? By the output from using the embeddings in a downstream model? Or is it that the floating-point values differ beyond some expected tolerance?

You are correct, we mean that the floating-point values differ beyond some expected tolerance.

In this table we compare ONNX embeddings in NodeJS (on the left) with ONNX embeddings in RN (on the right).
Given a query and list of sentences, we calculate the cosine similarity and sort the sentences to show the most similar first.
The ranked NodeJS sentences on the left feel right, while the RN ones are different. We expected they would be identical, and the table clearly shows how results are affected in RN.

<img width="730" alt="Screenshot 2024-02-05 at 16 11 07" src="https://github.com/microsoft/onnxruntime/assets/23238077/86fc4f93-e263-44ad-9685-39275952cc60">

**Comment by skottmckay** [comment]
I would suspect that python and nodejs are hitting the same low level code if the results are the same. 

As an additional data point, can you run your evaluation on another platform like x64 desktop or an actual iOS device instead of the simulator? Or alternatively you could enable the XNNPACK EP as an alternative implementation of MatMul on CPU (vs. ORT's MLAS library which is used by default)

**Comment by pax-k** [comment]
@skottmckay Thanks for the tip! 🙌 

Good news: if i use coreml (and not cpu) as an execution provider for ONNX in react-native iOS, then the ranked sentences match the nodejs ones! I also tested nnapi for Android, but it's as good as cpu.

Hopefully we can roll a text embedding model into production soon without issues.

Thank you for your patience! 🙏🏻

**Comment by skottmckay** [comment]
Excellent.

If you ever want to dig really really deeply into it you can do a custom build with a flag to output the result of individual nodes to compare platforms and see how things change throughout the model. 

https://onnxruntime.ai/docs/build/inferencing.html#debugnodeinputsoutputs

You could set ORT_DEBUG_NODE_IO_OP_TYPE_FILTER to limit to just the MatMul nodes.

**Comment by pax-k** [comment]
@skottmckay Could a custom build improve performance on iOS?

We did some benchmarks and seems that CoreML is slower than CPU on iOS (but CPU is not precise, so we don't use it). 

This is a benchmark for calculating the average inference time on-device with CoreML, using [Jina embeddings model](https://huggingface.co/Xenova/jina-embeddings-v2-small-en): 
```
 LOG  Model loaded successfully: jina-embeddings-v2-small-en
 LOG  Runs: 100
 LOG  Text size (chars): 1099
 LOG  Download time (s): 4.02
 LOG  Load time (s): 0.756
 LOG  Output dims: 512
 LOG  Average times (s): {
  "tokenize": 0.0025299999999999997,
  "inference": 0.21671
}
```

We get 220ms inference time on iOS + CoreML, compared to 5ms on NodeJS + CPU. We also tested other models, like bge-small-en-v1.5, gte-small, e5-small-v2, but jina-embeddings-v2-small-en was the fastest. 

Are you aware of any tweaks we could try to improve inference times?

Thanks!

**Comment by skottmckay** [comment]
Depends on the operations in the model. It will be slower if there are unsupported operators breaking up partitions between CoreML and CPU EP. Run with the log severity level set to 'VERBOSE' (0) in the session options and look for 'Node placements' in the output. 

FWIW I'm not convinced any particular EP is more 'precise'. I think they're all producing valid output, and saying one is more precise feels a little arbitrary and based on whether you liked the results for a specific query more or less for that EP. i.e. appointing the NodeJS output to be the measure of precision baseline may be flawed. if you run a wide range of queries the 'best' set of results for each may cycle between all the EPs you test with.
